### PR TITLE
IGNITE-20439 Sql. Support multiple schemas in CatalogSqlSchemaManager

### PR DESCRIPTION
**Pull request by korlov42**
Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/schema/SqlSchemaManagerImplTest.java [hunk 1]
```diff
@@ -0,0 +1,683 @@
...
+                Arguments.of(ColumnType.DECIMAL, 4, 0),
+                Arguments.of(ColumnType.DECIMAL, 4, 2),
+                Arguments.of(ColumnType.NUMBER, 4, -1),
+                Arguments.of(ColumnType.STRING, 40, -1),
+                Arguments.of(ColumnType.BYTE_ARRAY, 40, -1),
+                Arguments.of(ColumnType.DATE, -1, -1),
+                Arguments.of(ColumnType.TIME, 2, -1),
+                Arguments.of(ColumnType.DATETIME, 2, -1),
+                Arguments.of(ColumnType.TIMESTAMP, 2, -1),
+                Arguments.of(ColumnType.UUID, 2, -1),
```
Comment (lowka) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/schema/SqlSchemaManagerImplTest.java:592]: There was a typo in original code - UUID does not have precision, so it should be `Arguments.of(ColumnType.UUID, -1, -1),`.

**Comment by lowka** [comment]
Maybe we can need to add an end-to-end test (because I haven't found one) ? Something like this (it works):

>sql("CREATE TABLE known_views (id int, val VARCHAR, PRIMARY KEY (id) ) ");
>
>sql("INSERT INTO known_views VALUES(1, 'SYSTEM_VIEWS')");
>sql("INSERT INTO known_views VALUES(2, 'SYSTEM_VIEW_COLUMNS')");
>
>assertQuery("SELECT s.schema, s.name FROM PUBLIC.known_views, SYSTEM.SYSTEM_VIEWS as s " + 
>    "WHERE s.name = PUBLIC.known_views.val ORDER BY s.name")
>   .returns("SYSTEM", "SYSTEM_VIEW_COLUMNS")
>    .returns("SYSTEM", "SYSTEM_VIEWS")
>    .check();

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/schema/SqlSchemaManagerImplTest.java (hunk 1 above)
Comment (korlov42) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/schema/SqlSchemaManagerImplTest.java]: done

Comment (korlov42) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/schema/SqlSchemaManagerImplTest.java]: done

**Comment by korlov42** [comment]
> Maybe we can need to add an end-to-end test (because I haven't found one) ? Something like this (it works):

test added (see `ItSystemViewsTest#systemViewAndTableCrossQuery`)

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/schema/SqlSchemaManagerImpl.java [hunk 2]
```diff
@@ -98,7 +99,63 @@ public CompletableFuture<Void> schemaReadyFuture(int version) {
...
+
+                    assert subSchema != null : name;
+
+                    IgniteSchema schema = subSchema.unwrap(IgniteSchema.class);
+
+                    assert schema != null : "unknown schema " + subSchema;
+
+                    IgniteTable table = schema.getTable(tableId);
+
+                    if (table != null) {
```
Comment (xtern) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/schema/SqlSchemaManagerImpl.java:119]: Currently IgniteSchema#getTable has assertion such that table cannot be null here. I think this assertion should be reworked or removed.

Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/schema/SqlSchemaManagerImpl.java:119]: I totally missed that, thanks! 

fixed

Comment (xtern) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/schema/SqlSchemaManagerImpl.java:119]: It seems PlanningContext#schema should be fixed also to return the schema that is set in the session.

Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/schema/SqlSchemaManagerImpl.java:119]: this method is used during execution phase. Why did you mention PlanningContext? PlanningContext is used during planning (aka preparing, see PrepareService)

File: modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogTestUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+import org.hamcrest.Matcher;
+import org.junit.jupiter.api.Test;
+
+/**
+ * Tests to verify {@link CatalogTestUtils}.
+ */
+class CatalogTestUtilsTest {
+
+    /**
+     * Simple smoke test to verify test manager is able to process a few version of catalog
```
[the remark this answers is not in the document]
Comment (korlov42) [reply modules/catalog/src/test/java/org/apache/ignite/internal/catalog/CatalogTestUtilsTest.java]: fixed

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/schema/SqlSchemaManagerImpl.java (hunk 2 above)
Comment (xtern) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/schema/SqlSchemaManagerImpl.java:119]: Sorry, may bad, I thought that `config().getDefaultSchema();` returns PUBLIC schema always.

**Review by xtern — APPROVED** [review]

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/schema/SqlSchemaManager.java [hunk 4]
```diff
@@ -19,21 +19,29 @@
...
+     * Returns root schema by the given timestamp.
      */
-    SchemaPlus schema(@Nullable String name, long timestamp);
+    SchemaPlus schema(long timestamp);
+
+    /**
+     * Returns table by given id, which version correspond to the one from schema of given version.
+     *
+     * @param schemaVersion Version of the schema.
+     * @param tableId A identifier of a table of interest.
```
[the remark this answers is not in the document]
Comment (korlov42) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/schema/SqlSchemaManager.java]: fixed

**Pull request merged**